### ISSUE TITLE
Enhance drafter security

### ISSUE THREAD

**Issue by RickMoynihan** [Swirrl/drafter#425]
The recent potential exploit around https://github.com/Swirrl/drafter/issues/419 has highlighted how we should start doing something we've always intended, which is practicing "security in depth" with drafters security model.

This issue would have been far less dangerous if we:

1. Did not run stardog with `--disable-security` (this isn't quite as bad as it sounds because we don't expose stardog directly - but we should still do it)
2. More importantly enable [stardogs named graph security](https://www.stardog.com/docs/#_named_graph_security) to ensure that only specific kinds of user can see specific graphs.

This morning I had a think about how we might do this in practice, and thought I'd write up my findings here.  Essentially there are two broad approaches: 

- A. Map real users (i.e. every auth0 (or mongo) user) into a stardog user 1-1 (and keep them in sync)
- B. Map many admin users (many -> 1) to a single `pmdadmin` role in stardog, and map unauthenticated (public) users/ requests to a single `public` role in stardog.  Additionally define a stardog `system` role which drafter can itself can use to power some of its management ops.  Finally a `muttnik-downloads` user may be created with the `system` role too, in order to generate downloads in admin that bypass drafter etc.

Approach B is I think the preferrred option, as it increases our security in a way that involves a lot less work, than keeping the two systems in sync.  So this is the approach we assume here.  If we feel approach A is worth considering we can write up what is involved separately; however the trade off is essentially that approach B assumes admin users pose less of a threat to other admins than public users; whilst approach A is stricter in that it assumes other admins may be adversarial; however my feeling is that the extra complexity with A will mean it's significantly harder to build and ensure correctness.

## Steps to securing drafter/stardog in depth

1. Enable named graph security
2. In every PMD system create the following roles: `role:pmdadmin` `role:public` `role:system`.
3. In every PMD system create at least three users in each of those roles respectively lets call them `user:pmdadmin` `user:public` `user:system`.
4. Include the stardog client library dependency into drafter for managing roles only.  We could in principle also use it for querying via SPARQL; but for now I suggest we just use the standard RDF4j client without changing the implementation.
5. Create three "sparql repositories" configured with each of the three users.
6. Whenever drafter receives a HTTP request to run a query or data update operation we should via middlewares pick the appropriate repository.  For example an unauthenticated (public) query would run with `user:public` but an authenticated admin query would run via `role:pmdadmin`.
7. Whenever we create a draftgraph after first generating its UUID, we need to use the stardog API to add read and write permissions to the `role:pmdadmin`, i.e. via the commandline we would grant `read` and `write` to `pmd-database\http://drafter/blah/blah/<GUID>`.  NOTE stardog does *NOT* support wildcard syntax here :-(.
8. Similarly prior to each publish for each graph we are adding (and removing) we need to grant the `read` permission to the role `pmd:public`.
9. Additionally drafters operations itself will need to happen with a `drafter` user in the `system` role.  This will be required for all draftset operations, e.g. even just displaying the JSON metadata for a draftset; will need to query the drafter state graph and arbitrary other graphs.  NOTE this also means this role will need to maintain grant permissions on draft graph creation etc.
10. When we init the database we will need to construct these users and the appropriate drafter configs with passwords for the required users etc

If drafter and stardog were setup like this, then the exploit we recently patched would have been far less severe, as public users would not be able to query those graphs via `SERVICE <db://pmd-data> { ,,, }`.  Admin users would of course be able to do this, but as these users are typically known and employed by the government it is significantly less worrying than public users being able to do so.  Additionally admin users are of course prevented from doing this via the fix in #419.

TLDR; the point of doing this is to help mitigate security vulnerabilities we don't already know about.
